### PR TITLE
Add trace to SerializeObject

### PR DESCRIPTION
**Pull request by mborsz**
**What type of PR is this?**

/kind feature
**What this PR does / why we need it**:
It starts logging SerializeObject calls that takes more than 5s (i.e. most likely large lists or requests where the client is not receiving response fast enough).

The standard flow of SerializeObject call is:
* object serialization
* single deferredResponseWriter.Write call with entire object

by adding a trace steps to Write call, we should be able to distinguish time spent on in-memory serialization (kube-apiserver overload) and time used to write bytes to a client (caused by e.g. slow client).

Hopefully, this will help with https://github.com/kubernetes/kubernetes/issues/97798

**Which issue(s) this PR fixes**:

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

/assign @wojtek-t

### CODE REVIEW

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/responsewriters/writers.go [hunk 1]
```diff
@@ -177,9 +188,23 @@ type deferredResponseWriter struct {
...
 }
 
 func (w *deferredResponseWriter) Write(p []byte) (n int, err error) {
+	if w.trace != nil {
+		// This Step usually wraps in-memory object serialization.
+		w.trace.Step("About to start writing response", utiltrace.Field{"size", len(p)})
+
+		firstWrite := !w.hasWritten
+		defer func() {
+			w.trace.Step("Write call finished",
```
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/endpoints/handlers/responsewriters/writers.go:202]: How many Write() calls we can have per serialization?

I think protobuf encoder does O(1), but what about json/yaml? I would like to ensure that we can't end up with gazillions of those and we won't introduce visible overhead with those...

Comment (mborsz) [reply staging/src/k8s.io/apiserver/pkg/endpoints/handlers/responsewriters/writers.go:202]: For json/yaml it's a single call as well (see [1] for yaml, [2] for json) We already use assumption about single call for gzip compression (see comment for defaultGzipThresholdBytes constant)

[1] https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/runtime/serializer/json/json.go#L314
[2] https://golang.org/src/encoding/json/stream.go?s=5070:5117#L231 called here: https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/runtime/serializer/json/json.go#L327

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/responsewriters/writers.go [hunk 2]
```diff
@@ -86,11 +88,20 @@ func StreamObject(statusCode int, gv schema.GroupVersion, s runtime.NegotiatedSe
 // The context is optional and can be nil. This method will perform optional content compression if requested by
 // a client and the feature gate for APIResponseCompression is enabled.
 func SerializeObject(mediaType string, encoder runtime.Encoder, hw http.ResponseWriter, req *http.Request, statusCode int, object runtime.Object) {
+	trace := utiltrace.New("SerializeObject",
+		utiltrace.Field{"method", req.Method},
+		utiltrace.Field{"url", req.URL.Path},
```
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/endpoints/handlers/responsewriters/writers.go:93]: @liggitt @sttts @deads2k - just to be on the safe side - is the URL always set in non-test code?

**Review by wojtek-t — COMMENTED** [review]
This looks fine to - just waiting for final confirmation from apimachinery folks.

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/responsewriters/writers.go (hunk 2 above)
Comment (sttts) [reply staging/src/k8s.io/apiserver/pkg/endpoints/handlers/responsewriters/writers.go:93]: We do the same in k8s.io/apiserver/pkg/endpoints/handlers all over the place. So I think this is fine.

Comment (wojtek-t) [reply staging/src/k8s.io/apiserver/pkg/endpoints/handlers/responsewriters/writers.go:93]: Thanks!

**Pull request merged**